src/order/omega_complete_partial_order.lean

```diff
@@ -0,0 +1,524 @@
+
+lemma map_le_map {g : α →ₘ β} (h : f ≤ g) : c.map f ≤ c.map g :=
+λ i, by simp [mem_map_iff]; intros; existsi i; apply h
+
+/-- `chain.zip` pairs up the elements of two chains that have the same index -/
+@[simps {rhs_md := semireducible}]
+def zip (c₀ : chain α) (c₁ : chain β) : chain (α × β) :=
+preorder_hom.prod.zip c₀ c₁
+
+end chain
+
+end omega_complete_partial_order
+
+open omega_complete_partial_order
+
+section prio
+set_option default_priority 50 -- see Note [default priority]
+
+/-- Complete partial order (ωCPO) are useful for the formalization
+of the semantics of programming languages. Its notion of limit
+helps define the meaning of recursive procedures
+
+It has a supremum operation on increasing sequences indexed by natural
+numbers (which we call `ωSup`). In this sense, it is strictly weaker than
+join semi-lattice as only ω-sized totally ordered sets have a supremum. -/
+class omega_complete_partial_order (α : Type*) extends partial_order α :=
+(ωSup     : chain α → α)
+(le_ωSup  : ∀(c:chain α), ∀ i, c i ≤ ωSup c)
+(ωSup_le  : ∀(c:chain α) x, (∀ i, c i ≤ x) → ωSup c ≤ x)
+
+end prio
+
+namespace omega_complete_partial_order
+variables {α : Type u} {β : Type v} {γ : Type*}
+variables [omega_complete_partial_order α]
+
+/-- Transfer a `omega_complete_partial_order` on `β` to a `omega_complete_partial_order` on `α` using
+a strictly monotone function `f : β →ₘ α`, a definition of ωSup and a proof that `f` is continuous
+with regard to the provided `ωSup` and the ωCPO on `α`. -/
+@[reducible]
+protected def lift [partial_order β] (f : β →ₘ α)
+  (ωSup₀ : chain β → β)
+  (h : ∀ x y, f x ≤ f y → x ≤ y)
+  (h' : ∀ c, f (ωSup₀ c) = ωSup (c.map f)) : omega_complete_partial_order β :=
+{ ωSup := ωSup₀,
+  ωSup_le := λ c x hx, h _ _ (by rw h'; apply ωSup_le; intro; apply f.monotone (hx i)),
+  le_ωSup := λ c i, h _ _ (by rw h'; apply le_ωSup (c.map f)) }
+
+lemma le_ωSup_of_le {c : chain α} {x : α} (i : ℕ) (h : x ≤ c i) : x ≤ ωSup c :=
+le_trans h (le_ωSup c _)
+
+lemma ωSup_total {c : chain α} {x : α} (h : ∀ i, c i ≤ x ∨ x ≤ c i) : ωSup c ≤ x ∨ x ≤ ωSup c :=
+classical.by_cases
+  (assume : ∀ i, c i ≤ x, or.inl (ωSup_le _ _ this))
+  (assume : ¬ ∀ i, c i ≤ x,
+    have ∃ i, ¬ c i ≤ x,
+      by simp only [not_forall] at this ⊢; assumption,
+    let ⟨i, hx⟩ := this in
+    have x ≤ c i, from (h i).resolve_left hx,
+    or.inr $ le_ωSup_of_le _ this)
+
+lemma ωSup_le_ωSup_of_le {c₀ c₁ : chain α} (h : c₀ ≤ c₁) : ωSup c₀ ≤ ωSup c₁ :=
+ωSup_le _ _ $
+λ i, Exists.rec_on (h i) $
+λ j h, le_trans h (le_ωSup _ _)
+
+lemma ωSup_le_iff (c : chain α) (x : α) : ωSup c ≤ x ↔ (∀ i, c i ≤ x) :=
+begin
+  split; intros,
+  { transitivity ωSup c,
+    apply le_ωSup _ _, exact a },
+  apply ωSup_le _ _ a,
+end
+
+section continuity
+open chain
+
+variables [omega_complete_partial_order β]
+variables [omega_complete_partial_order γ]
+
+/-- a monotone function `f : α →ₘ β` is continuous if it distributes over ωSup -/
+def continuous (f : α →ₘ β) : Prop :=
```

Comment:
Should this be protected? The more familiar notion of `continuous` is in the root namespace.

Reply:
I don't see the need for this. It makes the use overly verbose and I don't see the two being used together.